EV3-G API GyroSensor

dwalton76 opened the pull request.

dwalton76 committed:
EV3-G API GyroSensor

dwalton76 commented:
For issue #362

dwalton76 commented:
This isn't ready yet, when I run the following:
```
robot@evb[bin]# cat gyro.py 
#!/usr/bin/env python3

from ev3dev.auto import GyroSensor
from pprint import pformat
import logging
import sys
import time


# logging
logging.basicConfig(level=logging.DEBUG,
                    format="%(asctime)s %(levelname)5s: %(message)s")
log = logging.getLogger(__name__)


gy = GyroSensor()
#gy.reset()

while True:
    print("angle %s" % gy.angle)
    print("rate %s" % gy.rate)
    (angle, rate) = gy.rate_and_angle
    print("rate %s, angle %s" % (rate, angle))
    time.sleep(1)
robot@evb[bin]# 
```

We get a TimeoutError when trying to set the mode
```
robot@evb[bin]# ./gyro.py 
Traceback (most recent call last):
  File "./gyro.py", line 20, in <module>
    print("angle %s" % gy.angle)
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post72-py3.5.egg/ev3dev/core.py", line 2622, in angle
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post72-py3.5.egg/ev3dev/core.py", line 2091, in mode
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post72-py3.5.egg/ev3dev/core.py", line 291, in set_attr_string
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post72-py3.5.egg/ev3dev/core.py", line 251, in _set_attribute
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post72-py3.5.egg/ev3dev/core.py", line 275, in _raise_friendly_access_error
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post72-py3.5.egg/ev3dev/core.py", line 247, in _set_attribute
TimeoutError: [Errno 110] Connection timed out
robot@evb[bin]# 
```

I haven't seen this issue on any other sensor....can someone with a gyro sensor try the patch in this PR and see if you get the same behavior?

WasabiFan commented on ev3dev/core.py:
```diff
@@ -2623,6 +2638,29 @@ def rate_and_angle(self):
+        return self.value(0)
+
+    @property
+    def direct(self):
```
Docs for `direct`?

dlech replied:
Should probably be `_direct`. I don't expect that it would be used normally, if at all.

dlech commented:
Check `dmesg`. It could be driver issues, e.g. https://github.com/ev3dev/ev3dev/issues/671

dwalton76 commented:
If I scroll back in dmesg a bit I see
```
[61668.755196] lego-port port5: Removed device 'evb-ports:in1:ev3-uart-host'
[61668.791500] lego-sensor sensor1: Unregistered 'lego-ev3-color' on 'evb-ports:in1'.
[61674.595972] lego-port port5: Added new device 'evb-ports:in1:ev3-uart-host'
[61675.805044] lego-sensor sensor3: Registered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62443.728409] lego-port port5: Removed device 'evb-ports:in1:ev3-uart-host'
[62443.758598] lego-sensor sensor3: Unregistered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62446.498956] lego-port port5: Added new device 'evb-ports:in1:ev3-uart-host'
[62447.703095] lego-sensor sensor4: Registered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62450.322760] tty ttyS4: Reconnected due to: Bad message size
[62858.070862] lego-port port5: Removed device 'evb-ports:in1:ev3-uart-host'
[62858.102750] lego-sensor sensor4: Unregistered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62859.440963] lego-port port5: Added new device 'evb-ports:in1:ev3-uart-host'
[62860.651662] lego-sensor sensor5: Registered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62953.283531] lego-port port5: Removed device 'evb-ports:in1:ev3-uart-host'
[62953.315735] lego-sensor sensor5: Unregistered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62954.734148] lego-port port5: Added new device 'evb-ports:in1:ev3-uart-host'
[62955.943074] lego-sensor sensor6: Registered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62958.562963] tty ttyS4: Reconnected due to: Bad message size
[62962.504001] lego-port port5: Removed device 'evb-ports:in1:ev3-uart-host'
[62962.530208] lego-sensor sensor6: Unregistered 'lego-ev3-gyro' on 'evb-ports:in1'.
[62965.264427] lego-port port5: Added new device 'evb-ports:in1:ev3-uart-host'
[62966.470619] lego-sensor sensor7: Registered 'lego-ev3-gyro' on 'evb-ports:in1'.
```

But when I unplug/plug the sensor I see
```
[99132.268283] lego-port port5: Removed device 'evb-ports:in1:ev3-uart-host'
[99132.300579] lego-sensor sensor7: Unregistered 'lego-ev3-gyro' on 'evb-ports:in1'.
[99133.758601] lego-port port5: Added new device 'evb-ports:in1:ev3-uart-host'
[99134.968876] lego-sensor sensor8: Registered 'lego-ev3-gyro' on 'evb-ports:in1'.
```

and then run
```
robot@evb[bin]# ./gyro.py 
angle 43
rate 0
rate 0, angle 0
Traceback (most recent call last):
  File "./gyro.py", line 20, in <module>
    print("angle %s" % gy.angle)
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post73-py3.5.egg/ev3dev/core.py", line 2622, in angle
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post73-py3.5.egg/ev3dev/core.py", line 2091, in mode
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post73-py3.5.egg/ev3dev/core.py", line 291, in set_attr_string
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post73-py3.5.egg/ev3dev/core.py", line 251, in _set_attribute
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post73-py3.5.egg/ev3dev/core.py", line 275, in _raise_friendly_access_error
  File "/usr/local/lib/python3.5/dist-packages/python_ev3dev-0.8.0.post73-py3.5.egg/ev3dev/core.py", line 247, in _set_attribute
TimeoutError: [Errno 110] Connection timed out
robot@evb[bin]# 
```
but I got no new output in dmesg when I ran gyro.py

dlech commented on ev3dev/core.py:
```diff
@@ -241,7 +241,11 @@ def _set_attribute(self, attribute, name, value):
+                    attribute.write(value)
+                else:
+                    attribute.write(value.encode())
+
```
You could drop the raw parameter and just say

```python
if isinstance(value, str):
    value = value.encode()
attribute.write(value)

dlech commented:
I'm not seeing a TimeoutError on EV3. Are you still running on EVB?

dwalton76 commented:
I'm seeing it on both.  My EV3 is running
```
robot@ev3alt[ev3dev-lang-python]# ev3dev-sysinfo 
Image file:         ev3dev-stretch-ev3-generic-2017-09-01
Kernel version:     4.9.49-ev3dev-1.5.1-ev3
Brickman:           0.10.0
ev3devKit:          0.5.2
Board:              board0
BOARD_INFO_HW_REV=7
BOARD_INFO_MODEL=LEGO MINDSTORMS EV3
BOARD_INFO_ROM_REV=6
BOARD_INFO_SERIAL_NUM=0016533F8FD1
BOARD_INFO_TYPE=main
robot@ev3alt[ev3dev-lang-python]# 
```

dwalton76 committed:
Drop 'raw' variable, minor cleanup

dwalton76 commented:
@dlech can you try a few different gyros and see if you see this with any of them? I think mine is an older one, maybe that is a factor.

dlech commented:
What is mfg code stamped on the sensor?

dlech commented:
Sure enough. I just got the TimeoutError with an 19N3 sensor (mfg 19th week of 2013)

dlech commented:
Would you mind open an issue for this? It's definitely a kernel driver issue. The code hasn't changed in 3 years, so I'm kind of surprised that no one has noticed before.

dwalton76 commented:
Mine is `03N3`

dwalton76 commented:
Opened https://github.com/ev3dev/ev3dev/issues/975

dwalton76 commented:
Any objections to merging this PR?  I'll test it more later once the kernel issue is fixed.
